sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ImplementSerializationMethodsCorrectly.cs

```diff
@@ -78,51 +84,46 @@ private static bool HasSerializationAttribute(IMethodSymbol methodSymbol)
                 methodSymbol.GetAttributes().Any(attr => attr.AttributeClass.IsAny(serializationAttributes));
         }
 
-        private static bool HasIssues(IMethodSymbol methodSymbol, out string errorMessage)
+        private static IEnumerable<string> FindIssues(IMethodSymbol methodSymbol)
         {
-            errorMessage = null;
-            var errors = new List<string>();
-
             if (methodSymbol.DeclaredAccessibility != Accessibility.Private)
             {
-                errors.Add(problemMakePrivateText);
+                yield return problemMakePrivateText;
             }
 
             if (!methodSymbol.ReturnsVoid)
             {
-                errors.Add(problemReturnVoidText);
+                yield return problemReturnVoidText;
             }
 
             if (!methodSymbol.TypeParameters.IsEmpty)
             {
-                errors.Add(problemGenericParameterText);
+                yield return problemGenericParameterText;
             }
 
             if (methodSymbol.Parameters.Length != 1 ||
                 !methodSymbol.Parameters.First().IsType(KnownType.System_Runtime_Serialization_StreamingContext))
             {
-                errors.Add(problemParameterText);
+                yield return problemParameterText;
             }
+        }
 
-            if (errors.Count > 0)
-            {
-                const string separator = ", ";
-                const string lastSeparator = " and ";
-
-                errorMessage = string.Join(separator, errors);
+        private static string BuildErrorMessage(IEnumerable<string> issues)
+        {
+            const string separator = ", ";
+            const string lastSeparator = " and ";
 
-                int lastCommaIdx = errorMessage.LastIndexOf(separator);
-                if (lastCommaIdx != -1)
-                {
-                    errorMessage = errorMessage
-                        .Remove(lastCommaIdx, separator.Length)
-                        .Insert(lastCommaIdx, lastSeparator);
-                }
+            string errorMessage = string.Join(separator, issues);
 
-                return true;
+            int lastCommaIdx = errorMessage.LastIndexOf(separator);
+            if (lastCommaIdx != -1)
```

Comment:
I don't know if this would be better but instead of joining all and replacing last if it exists I would have join all but last with comma and concat last with 'and'. WDYT?